quota/cache/cache.go

```diff
@@ -0,0 +1,200 @@
+		cache:      make(map[quota.Spec]*bucket),
+	}
+}
+
+func (m *manager) GetUser(ctx context.Context, req interface{}) string {
+	return m.qm.GetUser(ctx, req)
+}
+
+func (m *manager) PeekTokens(ctx context.Context, specs []quota.Spec) (map[quota.Spec]int, error) {
+	return m.qm.PeekTokens(ctx, specs)
+}
+
+func (m *manager) PutTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	return m.qm.PutTokens(ctx, numTokens, specs)
+}
+
+func (m *manager) ResetQuota(ctx context.Context, specs []quota.Spec) error {
+	return m.qm.ResetQuota(ctx, specs)
+}
+
+func (m *manager) GetTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Verify which buckets need more tokens, if any
+	specsToRefill := []quota.Spec{}
+	for _, spec := range specs {
+		bucket, ok := m.cache[spec]
+		if !ok || bucket.tokens < numTokens {
+			specsToRefill = append(specsToRefill, spec)
+		}
+	}
+
+	// Request the required number of tokens and add them to buckets
+	if len(specsToRefill) != 0 {
+		defer func() {
+			// Do not hold GetTokens on eviction, it won't change the result.
+			m.evictWg.Add(1)
+			go func() {
+				m.evict(ctx)
+				m.evictWg.Done()
+			}()
+		}()
+
+		// A more accurate count would be numTokens+m.batchSize-bucket.tokens, but that might force
+		// us to make a GetTokens call for each spec. A single call is likely to be more efficient.
+		tokens := numTokens + m.batchSize
+		if err := m.qm.GetTokens(ctx, tokens, specsToRefill); err != nil {
+			return err
+		}
+		for _, spec := range specsToRefill {
+			b, ok := m.cache[spec]
+			if !ok {
+				b = &bucket{}
+				m.cache[spec] = b
+			}
+			b.tokens += tokens
+		}
+	}
+
+	// Subtract tokens from cache
+	lastModified := now()
+	for _, spec := range specs {
+		bucket, ok := m.cache[spec]
+		// Sanity check
+		if !ok || bucket.tokens < 0 || bucket.tokens < numTokens {
+			glog.Warningf("Bucket invariants failed for spec %+v: ok = %v, bucket = %+v", spec, ok, bucket)
+			return nil // Something is wrong with the implementation, let requests go through.
+		}
+		bucket.tokens -= numTokens
+		bucket.lastModified = lastModified
+	}
+	return nil
+}
+
+func (m *manager) evict(ctx context.Context) {
+	m.mu.Lock()
+	// m.mu is explicitly unlocked, so we don't have to hold it while we wait for goroutines to
+	// complete.
+
+	if len(m.cache) <= m.maxEntries {
+		m.mu.Unlock()
+		return
+	}
+
+	// Find and evict the oldest entries. To avoid excessive token leakage, let's try and
+	// replenish the tokens held for the evicted entries.
+	var buckets bucketsByTime = make([]specBucket, 0, len(m.cache))
+	for spec, b := range m.cache {
+		if spec.Group != quota.Global {
+			buckets = append(buckets, specBucket{bucket: b, spec: spec})
+		}
+	}
+	sort.Sort(buckets)
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < len(m.cache)-m.maxEntries; i++ {
+		b := buckets[i]
+		glog.Infof("Evicting bucket for spec %+v (%v tokens held)", b.spec, b.tokens)
```

Comment:
This log message could probably be worded in a friendlier way for people not familiar with the quota code, e.g. 
```go
glog.V(1).Infof("Too many tokens cached - returning least recently used (%v tokens for %+v)", b.tokens, b.spec)
```